### PR TITLE
Forbid deep inheritance for the architecture classes

### PR DESCRIPTION
**Pull request by hackaugusto**
Nested inheritance is forbidden for architecture classes.

The design choices adopted by this architecture rely heavily on dynamic
dispatching based on type checking, e.g.:

```python
    if isintance(obj, StopStateChange):
        handle_stop(state, obj)
    elif isintance(obj, StopWithCareStateChange):
        handle_stop_with_care(state, obj)
```

For the above snippet to work, it's important for StopWithCareStateChange
to not be a subclass of StopStateChange, otherwise the order of the
statements used for the dynamic dispatch matter, leading to subtle bugs
like the one above.

### CODE REVIEW

File: raiden/transfer/architecture.py [hunk 1]
```diff
@@ -29,7 +29,54 @@
...
+        elif isintance(obj, StopWithCareStateChange):
+            handle_stop_with_care(state, obj)
+
+    For the above snippet to work, it's important for StopWithCareStateChange
+    to not be a subclass of StopStateChange, otherwise the order of the
+    statements used for the dynamic dispatch matter, leading to subtle bugs
+    like the one above.
+    """
+
+    def __new__(mcs, name, bases, namespace, **kwds):  # pylint: disable=unused-argument
```
Comment (LefterisJP) [inline raiden/transfer/architecture.py]: For consistency can you rename `**kwds` to `**kwargs`?

Also I have seen the `namespace` argument being called `attrs`. Can we also call it like that as I think it makes a bit more sense. It's the output of the `__prepare__` function right?

Note: on namespace naming not really strong opinion -- so if you prefer it like this I am fine.

File: raiden/transfer/architecture.py [hunk 2]
```diff
@@ -29,7 +29,54 @@
...
+            # Only one kind of marker may be used::
+            #
+            #      class Incorrect(State, StateChange): pass
+            #
+            # And a marker may be used at most once::
+            #
+            #      class Incorrect(State, State): pass
+            #
+            markers = [base for base in bases if isinstance(base, FlatInheritance)]
+            assert len(markers) == 1, FlatInheritance.__doc__
```
Comment (LefterisJP) [inline raiden/transfer/architecture.py]: Perhaps also add some title to the assertion error here? "Nested Inheritance Error"?
Same for the assertion below.

File: raiden/transfer/architecture.py [hunk 3]
```diff
@@ -29,7 +29,54 @@
...
+
+            # Enforce a flat inheritance for the marker::
+            #
+            #      class Correct(State): pass
+            #      class Incorrect(Correct): pass
+            marker = markers[0]
+            is_nested = any(isinstance(base, FlatInheritance) for base in marker.__bases__)
+            assert not is_nested, FlatInheritance.__doc__
+
+        return type.__new__(mcs, name, bases, dict(namespace))
```
Comment (LefterisJP) [inline raiden/transfer/architecture.py]: According to the article I read `namespace`, the value returned by `__prepare__` may not be a dict and we have to make sure it's a dict inside `__new__`. In our case here if the `dict()` constructor fails we just let the exception propagate I suppose?

**Comment by LefterisJP** [comment]
What you want to do makes total sense. So few lines of code to achieve it but so many new concepts .. at least for me.

If anyone else wants to read a nice article on those concepts I would recommend [this](https://blog.ionelmc.ro/2015/02/09/understanding-python-metaclasses/).

Note: A lot of stuff is happening under the hood with metaclasses.

**Comment by schmir** [comment]
The simplest solution, if you fear errors in your dispatching code, is not to
use isinstance checks. Just check if the type matches. The above dispatching
code would become:

```
   if type(obj) == StopStateChange:
       ...
   elif type(obj) == StopWithCareStateChange:
       ...
```

If you have a deep class tree, functools.singledispatch would also be an option,
which can do the dispatching for you.

So, let's assume there are other reasons to forbid inheriting from those classes.

This concept is known from java as 'final class', i.e. a class can be marked
final, which means you cannot subclass it. The PR you contributed is different 
in that it seems to allow one level of subclassing and not more [2].

If you really need that, I propose to use a solution that works like java final
classes. See [1] for an implementation.

Though some would consider such a solution as unpythonic. 


[1] https://stackoverflow.com/questions/3948964/in-python-how-can-i-prohibit-class-inheritance
[2] The 'final declaration' for a class is hidden in it's superclass in your implementation. If you
   really need final classes, please mark those classes instead of marking the
   superclass as 'this class can be subclassed, but only one level deep'.

**Comment by hackaugusto** [comment]
Yes, the motivation for this is exactly the to avoid problems with `isinstance`, the check using `type` is fine with me for that purpose. The concept `final` is a bit more strict, but it is fine to use it here too.

Which approach do you guys want to proceed with?

**Comment by schmir** [comment]
I'd just change the  dispatching code to use type instead of isinstance and be done with it. I don't see a reason to introduce 'final classes' or similar solutions.

**Comment by hackaugusto** [comment]
superseded by #1330

**Pull request closed without merging**